### ISSUE TITLE
Optimizers should use learning rates passed as tensors directly

### ISSUE THREAD

**Issue by johnc-keen** [pytorch/pytorch#106802]
### 🚀 The feature, motivation and pitch

If you pass a tensor as the learning rate when constructing an optimizer, it works, but internally it does an item() on every step, which breaks graph capture and triggers synchronization warnings if you have them enabled.

It would be nice if the optimizers (maybe just the fused versions?) directly used the tensor, because then you could adjust learning rate every step while using CUDA graphs.  I wind up forcing graph rebuilds every 1000 steps or so when I want to dynamically change learning rates.

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @ezyang

**Comment by janeyx99** [comment]
Hi, thanks for submitting this request! I have some questions to help clarify:
1. The ask is for lr to be treated as a scalar Tensor as there will be one lr for all the params at a time and NOT for a list of Tensors of lrs, yes? 
2. I think this is very relevant to the capturable implementation which is specifically meant for CUDA graphs. It would make total sense to have at least the capturable implementations accept Tensor scalars. Are you using optimizers with capturable already (e.g., Adam, AdamW, and as of yesterday NAdam)? If so, we can start with support there and then expand to other optimizers.
3. Just curious, are you using optimizers with fused=True? We have two general implementations (default=foreach, and fused) that should enable support for this and it'd be useful to know what to prioritize. For example, supporting foreach would require fixes that look like #106677. Supporting fused would require tweaking the kernels to accept Tensor lrs.

**Comment by vadimkantorov** [comment]
for foreach-ops, it might even make sense to accept all lr's as a tensor (an lr per param group) :)

**Comment by johnc-keen** [comment]
Yes, a single scalar value in a tensor is the ask. Per-layer lr for MuP and
so on would be valuable, but I don't see a super obvious way to transition
the dictionary based parameters to tensors, so that can be left for another
day.

Yes, I use capturable and fused flags for the optimizers.

Thank you!

**Comment by OrenLeung** [comment]
+1 this creates a graph break on torch.compile

we want to 2x speed up optimizer.step if this can be fixed.

In Jax, optimizer step, backwards, and forwards, grad update is just one graph. Would make sense to make this into one graph too in pytorch.

cc: @ezyang @Chillee @cchan @ezhang887

**Comment by ezyang** [comment]
cc @mlazos

**Comment by mlazos** [comment]
@janeyx99 can we make this the case for all optimizers - I also ran into this in internal use cases for Shampoo. 

The compiler supports this, it's just currently we generate guards on the LR since they are python scalars. This is also important for supporting cudagraphs across multiple iterations

**Comment by janeyx99** [comment]
Just an update: https://github.com/pytorch/pytorch/pull/106916 has been landed into the most recent nightly. People should now be able to use a tensor lr with the fused Adam or AdamW optimizers without any additional syncs between host and device. @johnc-keen let me know if you get a chance to try this out and whether it works as expected.

**Comment by johnc-keen** [comment]
I don't usually work with nightlies, so I probably won't test it until it
gets picked up by the Nvidia container, but I am looking forward to taking
advantage of it -- thank you!

**Comment by bobrenjc93** [comment]
Hi @janeyx99 are there any updates here? Should we close this out?

**Comment by janeyx99** [comment]
We can close this out as this has been working now!
